### ISSUE TITLE
Include generated autotools files in releases

### ISSUE THREAD

**Issue by orlitzky** [graph-algorithms/edge-addition-planarity-suite#5]
Hi, I package planarity for Gentoo linux and have a tiny problem. When you tag a release in github, it creates a "release tarball" from whatever files `git clone` would get you. That, however, is missing some important stuff. Most notably, it's missing the `./configure` script that people expect to use for the usual install method of `./configure && make && make install`.

Would you be willing to upload another tarball with all of those files to the release (3.0.1.0, and in the future)? The usual way of generating such a thing is

```
$ autoreconf -fi
$ ./configure
$ make dist
$ make distcheck (optional)
```

The last step only checks that the generated tarball will build once extracted. This will leave you with a file named `planarity-3.0.1.0.tar.gz`, in this case, that contains the configure script and other necessary bits and can be uploaded manually to the github "release" page.

We can work around this by requiring that everyone have autoconf, automake, etc. installed, but that slows things down a bit, and can cause unexpected compatibility problems between versions (autoconf-2.71 was just released after a decade, for example).

**Comment by graph-algorithms** [comment]
Hi Michael,

Thank you for your interest in this project. I have added Gentoo to the list of projects on the wiki page (https://github.com/graph-algorithms/edge-addition-planarity-suite/wiki). 

I'd like to make a change that will help resolve your issue, so let's collaborate on what to do exactly. 

I'd like to make a change that doesn't break other inclusions, such as Debian linux. The autogen.sh file was contributed by jdemeyer during inclusion in Debian linux, as I don't directly use autotools. Also, the source code tarball is generated automatically by the github 'Draft new release' functionality. So, I think the best approach is if you generate the desired config script and then send a PR so I can see what it says and where it goes and so I can ask @jdemeyer to have a look.

If all goes well, I'll accept the PR and then cut a 3.0.1.1 release that will include it. In the future, I'll use the pattern of what I see in the config script to amend it appropriately (if needed). Does this work for you?

Thanks,
John Boyer

**Comment by orlitzky** [comment]
It would work, but I don't want to give you bad advice. Adding the auto-generated files (there are more than just `configure`) to the repo would clutter it up and pollute your commit history with what is essentially junk. I know @jdemeyer &mdash; we are both (also) SageMath developers who are trying to improve the upstream software that, until now, has been bundled with SageMath. My ultimate goal here is just to help people install planarity on Gentoo, so that independent users of both planarity and SageMath don't have to install two separate copies of planarity (the Sage one and a standalone copy).

Some background about autotools: first, the current `autogen.sh` script is basically equivalent to the command `autoreconf -fi` that I mentioned. Both of them run a series of separate  commands &mdash; together called "autotools" &mdash;  in the correct order to generate a POSIX-compatible `configure` script and several `Makefile.in` files. (`configure.ac` generates `configure`, and `Makefile.am` generates `Makefile.in`, roughly). These are intended to be shipped to the user, who then runs `./configure`, which in turn converts each `Makefile.in` into a POSIX-compatible `Makefile`. Afterwards, he can build the software by running `make`.

The main benefits of this abstraction is that only developers need to have the autotools installed to generate the build script. End-users need only a POSIX-compatible shell (to run `./configure`) and a POSIX-compatible **Make** to run the `Makefile`s. But the other side of that is that the source repository is usually missing the "end user" stuff, which is what this issue is all about. Before Github, this was never really a problem because a "release" was usually an archive that somebody uploaded by hand to a website or FTP server. Since a github "release" is essentially just a zip file of a tag, an additional step is now needed. But the good news is that a lot of Github projects use autotools, so the workaround process is already standard: the developer runs `make dist`, and uploads the resulting tarball as an "asset" on the release page.

For an example, take the latest release of eclib: https://github.com/JohnCremona/eclib/releases/tag/20210503

The "Assets" listed there include the automatically-generated "Source code" archives, but also the `eclib-20210503.tar.gz` and `eclib-20210503.tar.bz2` that were made using `make dist`, and uploaded separately. Those contain the generated autotools stuff that the "Source code" archive is missing.

You _can_ commit those files to the repo so that they're included in the "Source code" archive, but it's a lot of stuff. Here's what I get on my machine after running `autoreconf -fi`:

```
mjo@gantu ~/src/edge-addition-planarity-suite $ autoreconf -fi
libtoolize: putting auxiliary files in '.'.
libtoolize: copying file './ltmain.sh'
libtoolize: Consider adding 'AC_CONFIG_MACRO_DIRS([m4])' to configure.ac,
libtoolize: and rerunning libtoolize and aclocal.
libtoolize: Consider adding '-I m4' to ACLOCAL_AMFLAGS in Makefile.am.
configure.ac:24: warning: The macro `AC_PROG_LIBTOOL' is obsolete.
configure.ac:24: You should run autoupdate.
aclocal.m4:121: AC_PROG_LIBTOOL is expanded from...
configure.ac:24: the top level
configure.ac:29: warning: AC_OUTPUT should be used without arguments.
configure.ac:29: You should run autoupdate.
configure.ac:23: installing './compile'
configure.ac:24: installing './config.guess'
configure.ac:24: installing './config.sub'
configure.ac:2: installing './install-sh'
configure.ac:2: installing './missing'
Makefile.am: installing './depcomp'

mjo@gantu ~/src/edge-addition-planarity-suite $ git status
On branch master
Your branch is up to date with 'origin/master'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	Makefile.in
	aclocal.m4
	autom4te.cache/
	compile
	config.guess
	config.sub
	configure
	depcomp
	install-sh
	ltmain.sh
	missing
```

Each of those files is unreadable by mere humans. The differences between them (when they change) are also uninteresting, since the changes are usually the result of some small version change in one of the autotools programs that generated them. As a result it's counterproductive to add them to the repo and track the changes to them; moreover, doing so will probably confuse anyone who is familiar with autotools.

In summary: yes, I can generate all of those files and add them to the repo, but you wouldn't be able to verify what any of them do, and it would make the repo harder to use. I don't recommend it. Another option would be for me to run `make dist` for you, and send you the resulting archive file to be uploaded to the "Releases" page. That would achieve the same end result as if you did it yourself, but you'd basically have to just trust me that the generated stuff doesn't contain anything malicious, because it's all incomprehensible. I trust me, but I don't recommend that approach _for you_ either, on principle. I'm certainly willing to do it though if that's what you prefer.

**Comment by graph-algorithms** [comment]
Hi Michael,

Thank you for the additional details. I've made the desired tarball for the 3.0.1.0 release, and it passes make distcheck on my end (under Windows). I'd like to ask you to double-check that the attached tarball is also good to go on your end. If so, then I should be able to edit the 3.0.1.0 release tag to add it there. I'll also modify the readme.md to state the make distribution process that will be supported going forward as well as the .gitignore. Your extra details above helped because you are getting one file ("compile") that I am not (and I am getting a couple others that are not showing up untracked for you). 
Anyway, please let me know if the attached works for you...
[planarity-3.0.1.0.tar.gz](https://github.com/graph-algorithms/edge-addition-planarity-suite/files/6673799/planarity-3.0.1.0.tar.gz)

Thanks again,
John Boyer

**Comment by orlitzky** [comment]
Awesome, thanks! I tried it "normally" with a `./configure && make && make install` on my laptop, and also as part of our Gentoo package. Both worked. Thanks again.

**Comment by graph-algorithms** [comment]
Hi Michael, 
Thank you for your help in making this project more consumable. I committed the gitignore and readme.md changes to close this issue, but I've now also edited the 3.0.1.0 release tag with the planarity-3.0.1.0.tar.gz distribution. Please let me know if it works as expected from that edited tag, or if anything else needs to be done.
Thank you again,
John Boyer

**Comment by orlitzky** [comment]
Everything is working as expected. We're using the github release for planarity-3.0.1.0 in Gentoo as of about an hour ago.
